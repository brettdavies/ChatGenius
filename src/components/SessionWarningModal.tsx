import { useAuth0 } from '@auth0/auth0-react';

interface SessionWarningModalProps {
  isOpen: boolean;
  remainingMinutes: number;
  onExtend: () => void;
  onClose: () => void;
}

const SessionWarningModal = ({
  isOpen,
  remainingMinutes,
  onExtend,
  onClose
}: SessionWarningModalProps) => {
  const { logout } = useAuth0();

  if (!isOpen) return null;

  return (
    <div 
      role="dialog"
      aria-labelledby="session-warning-title"
      aria-modal="true"
      className="fixed inset-0 bg-black bg-opacity-50 flex items-center justify-center z-50"
    >
      <div className="bg-white rounded-lg p-6 max-w-md w-full mx-4">
        <h2 id="session-warning-title" className="text-xl font-semibold mb-4">Session Expiring Soon</h2>
        <p className="mb-6">
          Your session will expire in approximately {remainingMinutes} minutes. 
          Would you like to extend your session?
        </p>
        <div className="flex justify-end space-x-4">
          <button
            onClick={() => {
              logout({ logoutParams: { returnTo: window.location.origin } });
              onClose();
            }}
            className="px-4 py-2 text-gray-600 hover:text-gray-800"
          >
            Logout Now
          </button>
          <button
            onClick={onExtend}
            className="px-4 py-2 bg-indigo-600 text-white rounded hover:bg-indigo-700"
          >
            Extend Session
          </button>
        </div>
      </div>
    </div>
  );
};

export default SessionWarningModal; 