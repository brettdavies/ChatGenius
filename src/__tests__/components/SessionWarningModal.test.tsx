import { render, screen, fireEvent } from '@testing-library/react';
import { useAuth0 } from '@auth0/auth0-react';
import SessionWarningModal from '../../components/SessionWarningModal';

// Mock useAuth0
jest.mock('@auth0/auth0-react');
const mockUseAuth0 = useAuth0 as jest.Mock;

describe('SessionWarningModal', () => {
  const mockOnExtend = jest.fn();
  const mockOnClose = jest.fn();
  const mockLogout = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    mockUseAuth0.mockReturnValue({ logout: mockLogout });
  });

  describe('snapshots', () => {
    it('matches snapshot when modal is closed', () => {
      const { container } = render(
        <SessionWarningModal
          isOpen={false}
          remainingMinutes={5}
          onExtend={mockOnExtend}
          onClose={mockOnClose}
        />
      );
      expect(container).toMatchSnapshot();
    });

    it('matches snapshot when modal is open', () => {
      const { container } = render(
        <SessionWarningModal
          isOpen={true}
          remainingMinutes={5}
          onExtend={mockOnExtend}
          onClose={mockOnClose}
        />
      );
      expect(container).toMatchSnapshot();
    });

    it('matches snapshot with different remaining minutes', () => {
      const { container } = render(
        <SessionWarningModal
          isOpen={true}
          remainingMinutes={1}
          onExtend={mockOnExtend}
          onClose={mockOnClose}
        />
      );
      expect(container).toMatchSnapshot();
    });
  });

  it('renders nothing when isOpen is false', () => {
    render(
      <SessionWarningModal
        isOpen={false}
        remainingMinutes={5}
        onExtend={mockOnExtend}
        onClose={mockOnClose}
      />
    );

    expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
  });

  it('renders modal content when isOpen is true', () => {
    const remainingMinutes = 5;
    
    render(
      <SessionWarningModal
        isOpen={true}
        remainingMinutes={remainingMinutes}
        onExtend={mockOnExtend}
        onClose={mockOnClose}
      />
    );

    expect(screen.getByRole('dialog')).toBeInTheDocument();
    expect(screen.getByText('Session Expiring Soon')).toBeInTheDocument();
    expect(screen.getByText(/Your session will expire in approximately/)).toBeInTheDocument();
    expect(screen.getByText('Extend Session')).toBeInTheDocument();
    expect(screen.getByText('Logout Now')).toBeInTheDocument();
  });

  it('calls onExtend when extend session button is clicked', () => {
    render(
      <SessionWarningModal
        isOpen={true}
        remainingMinutes={5}
        onExtend={mockOnExtend}
        onClose={mockOnClose}
      />
    );

    fireEvent.click(screen.getByText('Extend Session'));
    expect(mockOnExtend).toHaveBeenCalledTimes(1);
  });

  it('calls logout and onClose when logout button is clicked', () => {
    render(
      <SessionWarningModal
        isOpen={true}
        remainingMinutes={5}
        onExtend={mockOnExtend}
        onClose={mockOnClose}
      />
    );

    fireEvent.click(screen.getByText('Logout Now'));
    expect(mockLogout).toHaveBeenCalledWith({
      logoutParams: { returnTo: window.location.origin }
    });
    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });

  it('displays correct remaining minutes', () => {
    const remainingMinutes = 3;
    
    render(
      <SessionWarningModal
        isOpen={true}
        remainingMinutes={remainingMinutes}
        onExtend={mockOnExtend}
        onClose={mockOnClose}
      />
    );

    // Use a more flexible text matcher that accounts for whitespace and line breaks
    expect(screen.getByText((content) => {
      return content.includes(`${remainingMinutes} minutes`) &&
             content.includes('Your session will expire in approximately');
    })).toBeInTheDocument();
  });

  it('has correct ARIA attributes', () => {
    render(
      <SessionWarningModal
        isOpen={true}
        remainingMinutes={5}
        onExtend={mockOnExtend}
        onClose={mockOnClose}
      />
    );

    const dialog = screen.getByRole('dialog');
    expect(dialog).toHaveAttribute('aria-labelledby', 'session-warning-title');
    expect(dialog).toHaveAttribute('aria-modal', 'true');
    expect(screen.getByText('Session Expiring Soon')).toHaveAttribute('id', 'session-warning-title');
  });
}); 