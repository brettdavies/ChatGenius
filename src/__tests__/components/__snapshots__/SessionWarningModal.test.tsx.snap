// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SessionWarningModal snapshots matches snapshot when modal is closed 1`] = `<div />`;

exports[`SessionWarningModal snapshots matches snapshot when modal is open 1`] = `
<div>
  <div
    aria-labelledby="session-warning-title"
    aria-modal="true"
    class="fixed inset-0 bg-black bg-opacity-50 flex items-center justify-center z-50"
    role="dialog"
  >
    <div
      class="bg-white rounded-lg p-6 max-w-md w-full mx-4"
    >
      <h2
        class="text-xl font-semibold mb-4"
        id="session-warning-title"
      >
        Session Expiring Soon
      </h2>
      <p
        class="mb-6"
      >
        Your session will expire in approximately 
        5
         minutes. Would you like to extend your session?
      </p>
      <div
        class="flex justify-end space-x-4"
      >
        <button
          class="px-4 py-2 text-gray-600 hover:text-gray-800"
        >
          Logout Now
        </button>
        <button
          class="px-4 py-2 bg-indigo-600 text-white rounded hover:bg-indigo-700"
        >
          Extend Session
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`SessionWarningModal snapshots matches snapshot with different remaining minutes 1`] = `
<div>
  <div
    aria-labelledby="session-warning-title"
    aria-modal="true"
    class="fixed inset-0 bg-black bg-opacity-50 flex items-center justify-center z-50"
    role="dialog"
  >
    <div
      class="bg-white rounded-lg p-6 max-w-md w-full mx-4"
    >
      <h2
        class="text-xl font-semibold mb-4"
        id="session-warning-title"
      >
        Session Expiring Soon
      </h2>
      <p
        class="mb-6"
      >
        Your session will expire in approximately 
        1
         minutes. Would you like to extend your session?
      </p>
      <div
        class="flex justify-end space-x-4"
      >
        <button
          class="px-4 py-2 text-gray-600 hover:text-gray-800"
        >
          Logout Now
        </button>
        <button
          class="px-4 py-2 bg-indigo-600 text-white rounded hover:bg-indigo-700"
        >
          Extend Session
        </button>
      </div>
    </div>
  </div>
</div>
`;
